src/main/java/org/prebid/server/bidder/huaweiads/HuaweiUtilRequest.java

```diff
@@ -0,0 +1,474 @@
+                    .build();
+        }
+        return null;
+    }
+
+    private static HuaweiDevice resolveHuaweiDevice(BidRequest bidRequest, JacksonMapper mapper) {
+        final Device device = bidRequest.getDevice();
+        if (device != null) {
+
+            final ExtUserDataDeviceIdHuaweiAds data = resolveDeviceData(bidRequest, mapper);
+            final boolean dataNotNull = data != null;
+
+            final String oaid = dataNotNull ? resolveId(data.getOaid()) : null;
+            final String gaid = dataNotNull ? resolveId(data.getGaid()) : null;
+            final String imei = dataNotNull ? resolveId(data.getImei()) : null;
+            final String clientTime = dataNotNull ? resolveClientTime(data.getClientTime()) : null;
+
+            if (oaid == null && gaid == null && imei == null) {
+                throw new PreBidException("resolveHuaweiDevice: Imei, Oaid, Gaid are all empty or null");
+            }
+
+            final Integer dnt = device.getDnt();
+            final String country = resolveCountry(bidRequest);
+
+            return HuaweiDevice.builder()
+                    .type(device.getDevicetype())
+                    .useragent(device.getUa())
+                    .os(device.getOs())
+                    .version(device.getOsv())
+                    .maker(device.getMake())
+                    .model(StringUtils.isBlank(device.getModel())
+                            ? DEFAULT_MODEL_NAME
+                            : device.getModel())
+                    .height(device.getH())
+                    .width(device.getW())
+                    .language(device.getLanguage())
+                    .pxratio(device.getPxratio())
+                    .belongCountry(country)
+                    .localeCountry(country)
+                    .ip(device.getIp())
+                    .gaid(gaid)
+                    .oaid(oaid)
+                    .imei(imei)
+                    .clientTime(clientTime)
+                    .isTrackingEnabled(resolveIsTrackingEnabled(dnt, oaid))
+                    .isGaidTrackingEnabled(resolveIsTrackingEnabled(dnt, gaid)).build();
+        }
+        return HuaweiDevice.builder().build();
+    }
+
+    private static String resolveId(List<String> listId) {
+        if (listId == null) {
+            return null;
+        }
+        return !listId.isEmpty() ? listId.get(0) : null;
+    }
+
+    private static ExtUserDataDeviceIdHuaweiAds resolveDeviceData(BidRequest request, JacksonMapper mapper) {
+        final User user = request.getUser();
+        final ExtUserDataHuaweiAds extUserDataHuaweiAds;
+        if (user == null) {
+            return null;
+        }
+        ExtUser userExt = user.getExt();
+        if (userExt == null) {
+            return null;
+        }
+
+        try {
+            extUserDataHuaweiAds = mapper.mapper().readValue(userExt.getData().toString(),
+                    ExtUserDataHuaweiAds.class);
```

Comment:
can be simplified to **return mapper.mapper().readValue(userExt.getData().toString(), ExtUserDataHuaweiAds.class).getData();**